### ISSUE TITLE
NaNs and Infs in Spherical and Spins Transforms

### ISSUE THREAD

**Issue by SSL32081** [kazewong/jim#204]
When taking gradients of the two transforms:
`SphereSpinToCartesianSpinTransform` and `SpinAnglesToCartesianSpinTransform`
and evaluates at some particular values, the Jacobians returned are `NaNs` or `-Inf` depending on the case.

These happen usually when the desired vector to be transformed is aligned with the $z$-axis, or when they have null magnitudes.

The `-Inf` is not much of a problem from a theoretical standpoint; it occurs whenever some input parameters do not affect the transformed parameters, but coming back to this later.

The `NaNs`, however, are unexpected. They occur for two reasons:
1. gradient of `jnp.arctan2` at `(0, 0)` gives `NaN`
2. The gradient of this (↓) is `NaN` when `(x, y) = (0, 0)`
```
ArcCos[ z / Sqrt[ x² + y² + z² ] ]
```
(gradient refers to `jnp.jacfwd`)

In summary, the following are the behaviours of the current transforms:
For `SphereSpinToCartesianSpinTransform`:
**Forward case**
* `ln(|J|) = -inf` is when any of these is zeros:
  * `mag` , then the column of `theta, phi = 0.0`
  * `theta`  , then the column of `phi = 0.0`

**Backward case**
* `ln(|J|) = nan` is when any pair of these are zeros:
  * `x & y` (the z-axis), then the row of `tilt, phi = nan`

For `SpinAnglesToCartesianSpinTransform`:
**Forward case**
* `ln(|J|) = -inf` is when any of these are zeros:
  * `tilt_2`, then the column of `phi_12 = 0.0`
  * `a_1` → `tilt_1 = 0.0`
  * `a_2 ` → `tilt_2, phi_12 = 0.0`
* `ln(|J|) = nan` is when any pair of these are zeros:
  * `a_1 & a₂`  (Zero spin)
  * `tilt_1 & tilt_2`  (Aligned-spin)

**Backward case**
* `ln(|J|) = nan` is when any pair of these are zeros:
  * `s1_x & s1_y`, then the row of`tilt_1, phi_12 = nan`
  * `s2_x & s2_y` → `tilt_2, phi_12 = nan`

As an example, this is how the Jacobian matrix looks like these parameters:
```
# forward transform at these values:
{'a_1': 0.1, 'a_2': 0.1, 'tilt_1': 0.1, 'tilt_2': 0.0, 'phi_jl': 0.0, 'phi_12': 0.0, 
 'theta_jn': 0.7, 'M_c': 30.0, 'q': 0.9, 'phase_c': 0.1 }

      det(J): 0.000000e+00
 log(det(J)): -inf

            theta_jn      phi_jl       tilt_1       tilt_2       phi_12         a_1          a_2    
    iota   +1.000e+00   -2.621e-03   +8.131e-05   +6.617e-05   -0.000e+00   +7.943e-05   -1.734e-06  
    s1_x   -6.490e-06   -1.028e-03   -9.894e-02   +2.594e-05   +0.000e+00   -9.927e-02   -6.800e-07  
    s1_y   -6.271e-05   -9.930e-03   +1.055e-02   +2.507e-04   +0.000e+00   +1.058e-02   -6.571e-06  
    s1_z   +0.000e+00   +0.000e+00   -9.983e-03   -4.337e-20   +0.000e+00   +9.950e-01   +0.000e+00  
    s2_x   +3.956e-24   +6.264e-22   -1.943e-23   -9.947e-02   +0.000e+00   -1.898e-23   +6.053e-20  
    s2_y   +3.823e-23   +6.053e-21   -1.878e-22   +1.029e-02   +0.000e+00   -1.834e-22   -6.260e-21  
    s2_z   +0.000e+00   -7.898e-40   +6.200e-22   -2.567e-21   +0.000e+00   +4.043e-23   +1.000e+00
```

As mentioned before, the `Infs` are expected from the theoretical standpoint. These points also constitute a zero-measure set, which means practically, it is extremely unlikely to land on them during sampling. However, it may cause actual issues when one would like to sample the spin angles with a prior that is restricted to only aligned spins.

In summary, the occurrence of these `nan` and `inf` have well-understood reasons, and the other purpose of this issue is to also serve as a precaution and future reference when these corner cases show up in practice.
